examples/08_join_aggregation.py

```diff
@@ -1,302 +1,548 @@
+# negative predictions incur a negative gain.
 #
-# Our first step is to extract features from the timestamp, using the
-# |TableVectorizer|. Natively, it uses the |DatetimeEncoder| on datetime
-# columns, and doesn't interact with numerical columns.
-from skrub import DatetimeEncoder, TableVectorizer
+# Ultimately, we want to maximize this metric. To do so, we can train our learner to
+# minimize a proper scoring rule like the log loss.
+import sklearn
+from sklearn.metrics import log_loss, make_scorer
 
-table_vectorizer = TableVectorizer(datetime=DatetimeEncoder(add_weekday=True))
-X_date_encoded = table_vectorizer.fit_transform(X)
-X_date_encoded.head()
 
-###############################################################################
-# We can now make a couple of plots and gain some insight on our dataset.
-import seaborn as sns
-from matplotlib import pyplot as plt
+def credit_gain_score(y_true, y_pred, amount):
+    """Define our utility function.
 
-sns.set_style("darkgrid")
+    These numbers are entirely made-up, don't try this at home!
+    """
+    mask_tn = (y_true == 0) & (y_pred == 0)
+    mask_fp = (y_true == 0) & (y_pred == 1)
+    mask_fn = (y_true == 1) & (y_pred == 0)
 
+    # Refusing a fraud yields 0 €
+    fraudulent_refuse = 0
 
-def make_barplot(x, y, title):
-    fig, ax = plt.subplots(layout="constrained")
-    norm = plt.Normalize(y.min(), y.max())
-    cmap = plt.get_cmap("magma")
+    # Accepting a fraud costs its whole amount
+    fraudulent_accept = -amount[mask_fn].sum()
 
-    sns.barplot(x=x, y=y, palette=cmap(norm(y)), ax=ax)
-    ax.set_title(title)
-    ax.set_xticks(ax.get_xticks(), labels=ax.get_xticklabels(), rotation=30)
-    ax.set_ylabel(None)
+    # Refusing a legitimate basket transactions cost 5 €
+    legitimate_refuse = mask_fp.sum() * -5
 
+    # Accepting a legitimate basket transaction yields 7% of its amount
+    legitimate_accept = (amount[mask_tn] * 0.07).sum()
 
-# O is Monday, 6 is Sunday
+    return fraudulent_refuse + fraudulent_accept + legitimate_refuse + legitimate_accept
 
-daily_volume = X_date_encoded["timestamp_weekday"].value_counts().sort_index()
 
-make_barplot(
-    x=daily_volume.index,
-    y=daily_volume.values,
-    title="Daily volume of ratings",
-)
+def get_results(model, X_test, y_test, threshold, amount, time_to_fit):
+    y_proba = model.predict_proba(X_test)[:, 1]
+    return {
+        "log_loss": log_loss(y_test, y_proba),
+        "gain_score": credit_gain_score(y_test, y_proba > threshold, amount),
+        "y_proba": y_proba,
+        "y_test": y_test,
+        "time_to_fit": time_to_fit,
+    }
 
-###############################################################################
-# We also display the distribution of our target ``y``.
-rating_count = y.value_counts().sort_index()
 
-make_barplot(
-    x=rating_count.index,
-    y=rating_count.values,
-    title="Distribution of ratings given to movies",
-)
+sklearn.set_config(enable_metadata_routing=True)
```

Comment:
not necessarily important but I think this config will persist for subsequent examples